### PR TITLE
Use config binding generator in Microsoft.Extensions.Logging.Configuration assembly

### PR DESCRIPTION
**Pull request by layomia**
Contributes to unrooting `ConfigurationBinder` from the ASP.NET Goldilocks stage 2 showcase app - [TodosApi](https://github.com/aspnet/Benchmarks/tree/8c03b2411b78c691b00ac1610aac82a6d8849709/src/BenchmarksApps/TodosApi).

FYI @DamianEdwards @eerhardt.

### CODE REVIEW

**Comment by layomia** [comment]
@eerhardt @ericstj I know there's no functional difference, but do you reckon we should user the generator in place of https://github.com/dotnet/runtime/pull/82254 & https://github.com/dotnet/runtime/pull/82098? Upsides are having extra test coverage for the generator & that we wouldn't need custom logic if more config binding is needed in the future.

File: src/libraries/Microsoft.Extensions.Logging.Configuration/src/Microsoft.Extensions.Logging.Configuration.csproj [hunk 1]
```diff
@@ -36,4 +36,8 @@
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Options.ConfigurationExtensions\src\Microsoft.Extensions.Options.ConfigurationExtensions.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Configuration.Binder\gen\Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="true" />
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Configuration/src/Microsoft.Extensions.Logging.Configuration.csproj:40]: ><ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Configuration.Binder\gen\Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="true" /> [](http://example.com/codeflow?start=4&length=222)

does this mean we'll have the source gen packaged inside two packages now?

Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Configuration/src/Microsoft.Extensions.Logging.Configuration.csproj:40]: It's just meant to be consumed here, not redistributed.  When we redistribute a generator we'll use https://github.com/dotnet/runtime/blob/9d7c9d06f3e0a935ad7b68fee367e578b614c8aa/src/libraries/Microsoft.Extensions.Configuration.Binder/src/Microsoft.Extensions.Configuration.Binder.csproj#L35

@layomia can you double check that this doesn't make it into the nuspec as a dependency?  Just double check we have nothing mentioning Microsoft.Extensions.Configuration.Binder.SourceGeneration in the Microsoft.Extensions.Logging.Configuration package.

**Review by ericstj — APPROVED** [review]
Assuming this doesn't change the Microsoft.Extensions.Logging.Configuration package dependencies it's good to go :shipit:

File: src/libraries/Microsoft.Extensions.Logging.Configuration/src/Microsoft.Extensions.Logging.Configuration.csproj (hunk 1 above)
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging.Configuration/src/Microsoft.Extensions.Logging.Configuration.csproj:40]: Actually, test failure shows this is a problem:
```
 error NU1101: (NETCORE_ENGINEERING_TELEMETRY=Build) Unable to find package Microsoft.Extensions.Configuration.Binder.SourceGeneration. No packages exist with this id
```

Fix this by applying `PrivateAssets="All"` to the project reference.
```suggestion
    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Configuration.Binder\gen\Microsoft.Extensions.Configuration.Binder.SourceGeneration.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="true" PrivateAssets="All" />
```

**Comment by ericstj** [comment]
> but do you reckon we should use the generator in place of https://github.com/dotnet/runtime/pull/82254 & https://github.com/dotnet/runtime/pull/82098

As we discussed offline - yes this is a good idea.  You'll want to inspect the code diff when you do just to see that there's no significant perf or behavior diff in the generated code.  As we discussed - best to do that in a separate PR.

**Review by tarekgh — APPROVED** [review]

**Comment by layomia** [comment]
@dotnet-policy-service rerun

File: src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggerFilterConfigureOptions.cs [hunk 2]
```diff
@@ -52,8 +51,6 @@ private void LoadDefaultConfigValues(LoggerFilterOptions options)
                 }
             }
 
-            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
-                Justification = "IConfiguration.GetValue is safe when T is a bool.")]
             bool GetCaptureScopesValue(LoggerFilterOptions options) => _configuration.GetValue(nameof(options.CaptureScopes), options.CaptureScopes);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggerFilterConfigureOptions.cs:54]: This intermediate nested method can be removed now. It was only needed to put the UnconditionalSuppressMessage attribute on it.

**Comment by eerhardt** [comment]
We are seeing it in the charts as well:

image

**Pull request merged**